#%% [markdown]
# # Module 6 Problem Set
#
# Name: 
#
# This week's problem set will be short answer questions (no coding) that will assess your understanding of the concepts behind RNA-Seq preprocessing.
#
# You are welcome to reference the course materials and other external resources to help you with these questions.
#

#%% [markdown]
# ### 1. Quality Control:
# Why is read quality control an essential first step in RNA-Seq preprocessing? 
# 
# a)Describe at least three things that might contribute to lower quality reads in an RNA-seq dataset. 
# 
# b) List two commonly used tools for this process.

#%% [markdown]
# Answer:

#%% [markdown]
# ### 2. Adapter Trimming:
#
# a) Why is it necessary to trim adapter sequences from raw RNA-Seq reads?
#
# b) What potential issues might arise if adapter sequences are not properly removed?

#%% [markdown]
# Answer:

#%% [markdown]
# ### 3. Multi-mapping reads
# Given a bulk RNA-Seq dataset of sorted neurons, would you expect a higher percentage of multi-mapping reads compared to a non-neuronal population? Justify your answer.
#

#%% [markdown]
# Answer:

#%% [markdown]
# ### 4. Pseudoalignment vs. Alignment:
#
# a) Define pseudoalignment in the context of RNA-Seq.
#
# b) How does pseudoalignment differ from traditional alignment methods?
#
# c) Name a commonly used tool for each approach.

#%% [markdown]
# ### 5. Deciding on the Normalization Method:
# Considering the high heterogeneity of cell types in the brain and the differential levels of gene expression across cell types:
#
# a) Why is normalization crucial in neuronal RNA-Seq data analysis?
#
# b) Justify the choice of a particular normalization method, such as TPM (Transcripts Per Million), over basic read count or FPKM (Fragments Per Kilobase of transcript per Million mapped reads) for a study focusing on comparing gene expression levels across various neuronal cell types.

#%% [markdown]
# Answer:

#%% [markdown]
# ### 6. Long _and_ short reads
# 
# a) Given an RNA-Seq dataset that consists of both short Illumina reads and long Oxford Nanopore reads, how might your preprocessing steps differ between the two data types? 
#
# b) What challenges might arise when dealing with such a dataset?

#%% [markdown]
# Answer:

#%% [markdown]
# ### 7. Some RNA-Seq libraries are strand-specific, while others are not.
#
# a) Describe the main _conceptual_ difference between these two library preparations.
#
# b) Why might strand-specific libraries be particularly valuable in deciphering alternative splicing events in neurons?

#%% [markdown]
# Answer:

#%% [markdown]
# ### 8. Deciding on Depth:
#
# a) Explain the concept of sequencing depth in RNA-Seq.
#
# b) Imagine a scenario where deep sequencing of a neuronal sample might be crucial. Contrast this with a situation where shallow sequencing might suffice.

#%% [markdown]
# Answer:
